### ISSUE TITLE
Support xDS server subscription to client resources

### ISSUE THREAD

**Issue by htuch** [envoyproxy/envoy#11396]
In order to be able to gain knowledge of resource readiness (https://github.com/envoyproxy/envoy/issues/5174), a management server today needs to resort to OOB techniques to learn whether a listener (for example) has successfully been bound to a port. There are other examples of this, for example resources failing to warm due to dependency failure. It's not sufficient to rely on xDS's inbuilt ACK/NACK mechanism, since these events often occur long after `DiscoveryResponse` acknowledgement.

We've been brainstorming approaches here and one that seemed highly attractive in UDPA-TP discussions was to make xDS fully symmetric (similar to what MCP does), allowing the management server to subscribe to resources produced from the client. This reverse pub-sub would then allow for details that are today made available in configdump or CSDS to be supplied directly to a management server inline in an ADS stream.

Ideally we have a design doc to start this work that examines how we can retrofit this into v3 xDS so that we can gain benefit today.

**Comment by htuch** [comment]
CC @costinm @jpeach @mattklein123 @envoyproxy/api-shepherds

**Comment by markdroth** [comment]
I do like the idea of allowing the xDS protocol to be symmetric.  There are other use-cases, like LRS, that may also be addressed by this approach.

That having been said, I do wonder whether we can do something simpler for the specific problem of having the client report config health to the server.  It seems like we could solve that problem by extending the ACK/NACK mechanism built into the protocol.  For example, we could extend DiscoveryRequest to exlicitly indicate (a) the config version currently being used, (b) the config version currently being warmed up, and (c) the last config version that attempted to warm and was unsuccessful.

**Comment by htuch** [comment]
This is possible, the main counter-argument being I think that this is going to increase the complexity of config version handling in xDS, and it's already pretty complicated.

**Comment by markdroth** [comment]
It seems like most of the complexity in the current mechanism is due to the strange way that version and nonce interact in requests.  I don't think there's actually anything inherently complicated about this problem, and I think there are other ways of structuring it that would be very straightforward.  If we just explicitly indicated the state of a given version in the request, then the need for nonce would basically go away, and I think this would get a lot simpler.

**Comment by htuch** [comment]
One of the problems you will run into with this is that there isn't a "config" being warmed. Once a config is accepted, you have a bunch of independent listeners or clusters being warmed. There would need to be some additional state tracking to make this work, and it might not be what the server actually wants. It wants to know whether listener X is active, not necessarily the details of X's earlier configuration.

It's a reasonable alternative to consider, but we should do some due diligence to find the most effective way to do this in v3 generally with minimal complexity.

**Comment by markdroth** [comment]
That's a fair point.  So perhaps what we need is to report state on a per-resource basis, rather than a per-response basis.  And once we've done that, it might not actually be simpler than having a separate resource published to indicate state.

I guess let's wait to have a concrete proposal and then we can debate the details.  Thanks!

**Comment by costinm** [comment]
And ideally an incremental approach/implementation: we can start with just adding a Resources[] entry to DiscoveryRequest.

Maybe also expose a way for WASM/modules to initiate sending of such DiscoveryRequest and getting back the associated DiscoveryResponse.

**Comment by htuch** [comment]
Also related to #9447

**Comment by markdroth** [comment]
I plan to work on a design for this later this quarter.

**Comment by kyessenov** [comment]
@markdroth Did we ever come up with a design for the reverse xDS?

**Comment by markdroth** [comment]
No, we haven't come up with a design proposal yet.  I do still intend to get back to this at some point, but it hasn't bubbled up to the top of the priority list yet. :(

The highest priority problem here is providing a better observability story for cases where the xDS resources are individually valid (and therefore get ACKed) but where a problem may show up when actually trying to apply a set of resources as a complete configuration.  One thing I've been thinking about here is whether that use-case is best served by making xDS symmetrical or whether it would make more sense to leverage metric exports via something like OpenTelemetry instead.

Even if we wind up deciding that that particular use-case is better served via metric exports, there may still be other use-cases that would benefit from making xDS symmetrical, such as avoiding the need for a separate stream for LRS.  But that would probably be a lower priority problem.

**Comment by kyessenov** [comment]
I don't think metric export is enough IMO. We are looking to broad cast events (e.g. a listener is warmed and ready), and it maps better to an event log entry rather an aggregated metric. A metric for this purpose is a poor fit, you'd be imposing a cost to aggregate this event metrics for each listener name (high cardinality), with unclear value of the measurement (e.g. time from ACK to warmed).

But I do think Otel log record might be a good fit. The only drawback is decoupling from xDS, as this would be essentially application-specific logs streamed outside, but we could loosely align on conventional xDS attributes to represent the entry uniformly across xDS clients.

**Comment by markdroth** [comment]
If what you're looking for is log entries, then I don't think that's a very good fit for xDS.  The xDS transport protocol is essentially a pub/sub mechanism: the idea would be that the server would subscribe to a resource and the client would send updates any time that resource changes.  But a log entry isn't really a new value for a resource; it's more of an append-to-log thing, which isn't a good fit for the xDS transport protocol design.  You might be able to shoehorn this into a pub/sub model, where the server side is just expected to know that every resource update is really a new log message, not a replacement for the log message it had previously, but that seems like a very confusing use of xDS.

I think OT logging would be a much better fit than xDS for what you describe.
